src/distributions/adoptium/installer.ts

```diff
@@ -0,0 +1,146 @@
+      const availableOptionsMessage = availableOptions
+        ? `\nAvailable versions: ${availableOptions}`
+        : '';
+      throw new Error(
+        `Could not find satisfied version for SemVer '${version.raw}'. ${availableOptionsMessage}`
+      );
+    }
+
+    return resolvedFullVersion;
+  }
+
+  protected async downloadTool(javaRelease: JavaDownloadRelease): Promise<JavaInstallerResults> {
+    let javaPath: string;
+    let extractedJavaPath: string;
+
+    core.info(
+      `Downloading Java ${javaRelease.version} (${this.distribution}) from ${javaRelease.url} ...`
+    );
+    const javaArchivePath = await tc.downloadTool(javaRelease.url);
+
+    core.info(`Extracting Java archive...`);
+    let extension = getDownloadArchiveExtension();
+
+    extractedJavaPath = await extractJdkFile(javaArchivePath, extension);
+
+    const archiveName = fs.readdirSync(extractedJavaPath)[0];
+    const archivePath = path.join(extractedJavaPath, archiveName);
+    const version = this.getToolcacheVersionName(javaRelease.version);
+
+    javaPath = await tc.cacheDir(archivePath, this.toolcacheFolderName, version, this.architecture);
+
+    if (process.platform === 'darwin') {
+      javaPath = path.join(javaPath, MACOS_JAVA_CONTENT_POSTFIX);
+    }
+
+    return { version: javaRelease.version, path: javaPath };
+  }
+
+  private async getAvailableVersions(): Promise<IAdoptiumAvailableVersions[]> {
+    const platform = this.getPlatformOption();
+    const arch = this.architecture;
+    const imageType = this.packageType;
+    const versionRange = '[1.0,100.0]'; // retrieve all available versions
+    const encodedVersionRange = encodeURI(versionRange);
+    const releaseType = this.stable ? 'ga' : 'ea';
+
+    console.time('adopt-retrieve-available-versions');
+
+    const baseRequestArguments = [
+      `project=jdk`,
+      'vendor=adoptopenjdk',
+      `heap_size=normal`,
+      `jvm_impl=hotspot`,
+      'sort_method=DEFAULT',
+      'sort_order=DESC',
+      `os=${platform}`,
+      `architecture=${arch}`,
+      `image_type=${imageType}`,
+      `release_type=${releaseType}`
+    ].join('&');
+
+    // need to iterate through all pages to retrieve the list of all versions
+    // Adopt API doesn't provide way to retrieve the count of pages to iterate so infinity loop
+    let page_index = 0;
+    const availableVersions: IAdoptiumAvailableVersions[] = [];
+    while (true) {
+      const requestArguments = `${baseRequestArguments}&page_size=20&page=${page_index}`;
+      const availableVersionsUrl = `https://api.adoptopenjdk.net/v3/assets/version/${encodedVersionRange}?${requestArguments}`;
+      if (core.isDebug() && page_index === 0) {
+        // url is identical except page_index so print it once for debug
+        core.debug(`Gathering available versions from '${availableVersionsUrl}'`);
+      }
+
+      const paginationPage = (
+        await this.http.getJson<IAdoptiumAvailableVersions[]>(availableVersionsUrl)
+      ).result;
```

Comment:
I think http-client will report error message and status code in case of failure (if statusCode is not successful).
Also I have checked Adopt swagger API, http-client should handle all status codes properly (the same for Azul)

Currently, we don't have retry logic and try-catch on task level and I am not sure that it makes sense to introduce it.